README.md
<p align="center"><img alt="logo" src="res/logo.png" width="200"></p>

<h1 align="center">HANA</h1>

[![Build Status](https://github.com/HancomAC/HANA/actions/workflows/.github/workflows/ci.yml/badge.svg)](https://github.com/HancomAC/HANA/actions/workflows/ci.yml)
[![CodqQL Analysis Status](https://github.com/HancomAC/HANA/actions/workflows/.github/workflows/codeql.yml/badge.svg)](https://github.com/HancomAC/HANA/actions/workflows/codeql.yml)
[![Docker Image Size (tag)](https://img.shields.io/docker/image-size/hancomac/hana/lastest)](https://hub.docker.com/r/hancomac/hana)

HANA is a PS Judgement Server that runs on Docker.

## Supported Languages

C, C++, Go, Java, Python3, Pypy3, Rust, Text, JavaScript, TypeScript

<details>
<summary>Build option</summary>

-   C

```shell
gcc Main.c -o Main -O2 -Wall -lm --static -std=c99 -DONLINE_JUDGE
```

-   C++

```shell
g++ Main.cpp -o Main -O2 -Wall -lm --static -pipe -std=c++17 -DONLINE_JUDGE
```

-   Go

```shell
go build Main.go
```

-   Java

```shell
javac --release 11 -J-Xms1024m -J-Xmx1920m -J-Xss512m -encoding UTF-8 Main.java
```

-   Python3

```shell
python3 -m compileall -b ${path}
```

-   Pypy3

```shell
pypy3 -m compileall -b ${path}
```

-   Rust

```shell
rustc Main.rs
```

-   TypeScript

```shell
tsc Main.ts
```

</details>

## Installation

```shell
docker run -p 80:80 -it hancomac/hana:lastest
```

## Usage

### Submit

Submit judge request can be done by following request.

You should use `POST` method and `application/json` as content type.

#### API Endpoint

```
/judge
```

#### Body

```typescript
interface Request {
    uid: string
    timeLimit: number
    memoryLimit: number
    language:
        | 'TEXT'
        | 'C'
        | 'CPP'
        | 'JAVA'
        | 'PYTHON3'
        | 'PYPY3'
        | 'RUST'
        | 'JAVASCRIPT'
        | 'TYPESCRIPT'
        | 'GO'
    judgeType:
        | 'CommonJudge'
        | 'OutputOnly'
        | 'Interactive'
        | 'SpecialJudge'
        | 'COMPETITIVE'
    source: {
        name: string
        source: string
    }[]
    dataSet: {
        scoringType: 'PROPORTIONAL' | 'QUANTIZED'
        data: {
            input?: string
            output: string
        }[]
    }[]
    specialJudge?: {
        language:
            | 'TEXT'
            | 'C'
            | 'CPP'
            | 'JAVA'
            | 'PYTHON3'
            | 'PYPY3'
            | 'RUST'
            | 'JAVASCRIPT'
            | 'TYPESCRIPT'
            | 'GO'
        source: string
    }
}
```

##### Example

```json
{
    "language": "CPP",
    "judgeType": "SpecialJudge",
    "specialJudge": {
        "language": "CPP",
        "source": "#include <stdio.h>\n#include <string.h>\n\nFILE * fin;\nFILE * fsol;\nFILE * fuser;\n\nint main(int argc,char *args[])\n{\n\tfin=fopen(args[1],\"r\");\n\tfsol=fopen(args[2],\"r\");\n\tfuser=fopen(args[3],\"r\");\n\n\tint ret=0;\n\tchar a[100], c[100];\n\twhile(fscanf(fsol,\" %[^\\r\\n]s\",a) != EOF){\n  fscanf(fuser,\" %[^\\r\\n]s\",c);\n  if(strcmp(a,c)!=0) {\n  ret=1;\n  break;\n  }\n\t}\n\tfclose(fin);\n\tfclose(fsol);\n\tfclose(fuser);\n\treturn ret; // 0 이면 성공, 1이면 실패\n}\n"
    },
    "source": [
        {
            "name": "Main.cpp",
            "source": "#include <iostream>\nusing namespace std;\nint main(){\n    int a,b;\n    while(cin >> a >> b)\n        cout << a+b << endl;\n return 0;\n}"
        }
    ],
    "dataSet": [
        {
            "scoringType": "QUANTIZED",
            "data": [
                { "input": "1 1", "output": "2" },
                { "input": "10 1", "output": "11" },
                { "input": "5 1", "output": "6" },
                { "input": "1 3", "output": "4" }
            ]
        },
        {
            "scoringType": "QUANTIZED",
            "data": [
                { "input": "15 1", "output": "16" },
                { "input": "17 21", "output": "38" }
            ]
        },
        {
            "scoringType": "QUANTIZED",
            "data": [
                { "input": "1 1", "output": "2" },
                { "input": "10 1", "output": "11" },
                { "input": "5 1", "output": "6" },
                { "input": "1 3", "output": "4" }
            ]
        }
    ],
    "timeLimit": 2000,
    "memoryLimit": 1024
}
```

#### Response

```typescript
interface Response {
    success: boolean
    uid: string
}
```

uid is an unique id for each judge request.

### Getting Response

There are two ways to retrieve result from HANA: Using WebSocket or using HTTP Polling.

#### Using WebSocket

You need to connect to `/` endpoint.

Response will be sent by WebSocket message.

#### Using HTTP Polling

You need to request to `/pull` endpoint with GET method.

Response will be sent by JSON, in `data` field as array.

##### Response Format

```typescript
interface responseOne {
    success: boolean
    type:
        | 'JUDGE_FINISH'
        | 'JUDGE_PROGRESS'
        | 'JUDGE_INFO'
        | 'JUDGE_STATUS'
        | 'JUDGE_ERROR'
    data?: any
    error?: string
}
```

##### Response Type

-   JUDGE_PROGRESS

Message will be sent when judge is progressing.

```typescript
interface responseFinish extends responseOne {
    data: {
        uid: string
        progress: number
        resultCode: 'RUN'
    }
}
```

progress is value between 0 and 1.

-   JUDGE_FINISH

Message will be sent when judge is finished.

```typescript
interface responseFinish extends responseOne {
    data: {
        uid: string
        result: (number | number[])[]
        resultCode: 'AC' | 'WA' | 'TLE' | 'MLE' | 'OLE' | 'RE' | 'CE'
        reason: ('AC' | 'WA' | 'TLE' | 'MLE' | 'OLE' | 'RE' | 'CE')[]
        time: number[]
        memory: number[]
        message: string
    }
}
```

result for each subtask will be sent as array.
element of result will be number if scoringType is QUANTIZED, or number[] if scoringType is PROPORTIONAL.

resultCode is a representing result of each subtask's reason.

message is one sample stderr message when compile error or runtime error.

-   JUDGE_INFO

Message will be sent when user requests to get judge program's info.

-   JUDGE_STATUS

Message will be sent when user requests to get judge status.

-   JUDGE_ERROR

Message will be sent when judge encounters error.

## Test Page

![test page](res/test_1.png)

You can test this program by opening `/test` endpoint in your browser.

You can judge 'A+B' problem with sample data.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.4](https://github.com/HancomAC/HANA/compare/v0.0.3...v0.0.4) (2022-01-13)

### Bug Fixes

-   replace express with koa ([d3fa6e1](https://github.com/HancomAC/HANA/commit/d3fa6e1c1204e96c4ba873321fd8cf94f0711df1))
-   update packages ([3218933](https://github.com/HancomAC/HANA/commit/3218933db954e2eaf495fb52a8219b37c90c5142))

### [0.0.3](https://github.com/HancomAC/HANA/compare/v0.0.2...v0.0.3) (2022-01-09)

### Features

-   automatically set config based on system performance ([23c6f76](https://github.com/HancomAC/HANA/commit/23c6f76ff4501fb4984e250fbcf9a7d4782a42e6))

### Bug Fixes

-   accept big input ([cf2beef](https://github.com/HancomAC/HANA/commit/cf2beefe400eb758ade27249c22632ef16b7308e))
-   fixed bug that not judging without SPJ ([9dd9677](https://github.com/HancomAC/HANA/commit/9dd9677fb5d469e03f97001522fbc14671e4487b))
-   ready for c# support ([5edefcb](https://github.com/HancomAC/HANA/commit/5edefcb7c4990e239dd7a91099a5b2c589871ea7))
-   reduce image size ([e9df2ec](https://github.com/HancomAC/HANA/commit/e9df2ecc94a676bfe87c8f72d89e3e29c26e11ef))

### [0.0.2](https://github.com/HancomAC/HANA/compare/v0.0.1...v0.0.2) (2022-01-08)

### Features

-   add http pull feature ([17abb15](https://github.com/HancomAC/HANA/commit/17abb1548f1755515e84ad692fcdcf254eb2128d))
-   add php support ([0ac32b0](https://github.com/HancomAC/HANA/commit/0ac32b09ba748f4714d17451ea81e1befb8c4879))
-   add ruby support ([c437724](https://github.com/HancomAC/HANA/commit/c4377244e4ef08148c9e22c8babde7687103620f))
-   make main process to not die ([961d11c](https://github.com/HancomAC/HANA/commit/961d11c820dddda8495d8dec20b6a9d4598ff27a))

### Bug Fixes

-   assert type of request ([284c42e](https://github.com/HancomAC/HANA/commit/284c42edd3adf929873f88450716e5462be56154))
-   assert type of request ([0381b9d](https://github.com/HancomAC/HANA/commit/0381b9dbd99531db529f5bef893c3594d32e535b))
-   kill zombie process ([007a0c8](https://github.com/HancomAC/HANA/commit/007a0c83672e943e53a46ae50aaf595a73807e3c))
-   kill zombie process more recursively ([87480b4](https://github.com/HancomAC/HANA/commit/87480b495b06755269fb765f32f49499d35ea888))
-   ready for c# support ([b162675](https://github.com/HancomAC/HANA/commit/b162675a21587d48032e5d88308fb34c7c37f999))

### 0.0.1 (2022-01-04)

### Features

-   add config file ([e3cf2f5](https://github.com/HancomAC/HANA/commit/e3cf2f56bbcc7216e736f33e09f898fa231cd2c1))
-   add example code ([af32d51](https://github.com/HancomAC/HANA/commit/af32d516d3e74bc02972c33514e2cc6c2eea195e))
-   add experimental output only judge ([19de162](https://github.com/HancomAC/HANA/commit/19de1626a18f33359afec5479a4570a51fb1a5d5))
-   add experimental python3 & pypy3 judge ([8fe8abf](https://github.com/HancomAC/HANA/commit/8fe8abf606cce7b2e4f7d7bc32705dba94e867bd))
-   add multi processing judge feature ([cedcdc5](https://github.com/HancomAC/HANA/commit/cedcdc569307cd1c7bac316d50b6397d2545e5c2))
-   add numpy & pandas support for python ([8cf2baf](https://github.com/HancomAC/HANA/commit/8cf2baf16cf4ab2998da09db62f388f860bd01a2))
-   add socket request feature ([9da8dfe](https://github.com/HancomAC/HANA/commit/9da8dfe7951a8cf1e559c5457ff25d9e97cbbb8b))
-   add special judge ([f42f790](https://github.com/HancomAC/HANA/commit/f42f790ecaea93d57289a95992d658bdccf16180))
-   add support for golang & rust ([d809e06](https://github.com/HancomAC/HANA/commit/d809e06791c6b6c8844b103a8323b4be7f955174))
-   add support for java ([ecc8fba](https://github.com/HancomAC/HANA/commit/ecc8fba6dcdd956d848e5281f1a4ec333ea06465))
-   add support for javascript & typescript ([d34b105](https://github.com/HancomAC/HANA/commit/d34b10564cee569f6fafcc4d4a918f6a13413fd9))
-   asynchronously build source file ([fd19d0c](https://github.com/HancomAC/HANA/commit/fd19d0c4237022e449d073d1729e595566792202))
-   build cpp file ([cc553b1](https://github.com/HancomAC/HANA/commit/cc553b18e31db6cead36d053ccab99c2329eb207))
-   enhance test page ([b949038](https://github.com/HancomAC/HANA/commit/b949038ebdfc2cbf38b4536a302d6a4d72ad4bcc))
-   enhance UI ([cadf2c5](https://github.com/HancomAC/HANA/commit/cadf2c5b45cda85c56c74ad4a10784d34e87d6f8))
-   judge cpp file experimentally ([201bc2d](https://github.com/HancomAC/HANA/commit/201bc2d84b36c36d38238de285d9c5c0a471e540))
-   judge one by one ([7c7e294](https://github.com/HancomAC/HANA/commit/7c7e29493ccde95c742aa4bdecfc5fa5f0c3297d))
-   keep message on sent fail ([6ade953](https://github.com/HancomAC/HANA/commit/6ade9531ecc3a56236dc10b0e64d42315a579d73))
-   ready for add rust & java support ([3a44836](https://github.com/HancomAC/HANA/commit/3a448362922cf36987b1b93896ed5db999a3ce1d))
-   ready for memory & time limit customizing ([ddda02b](https://github.com/HancomAC/HANA/commit/ddda02b0596ad4b167ec5136a68a451202b3a290))
-   support c language ([d7c6b40](https://github.com/HancomAC/HANA/commit/d7c6b408f08b1674968c4ffcf93fc3045444ea88))
-   support subtask judge ([682e352](https://github.com/HancomAC/HANA/commit/682e352d182abbc8d8987132e8cb14c17a4c6451))

### Bug Fixes

-   add source type ([3cee25a](https://github.com/HancomAC/HANA/commit/3cee25a8aab9312f2ca34459412295cccf6d4c05))
-   automatically trim output and compare ([0a3e9c3](https://github.com/HancomAC/HANA/commit/0a3e9c3f92b1f146fbec7957380b378443263515))
-   build source with user permission ([5461f91](https://github.com/HancomAC/HANA/commit/5461f91a393e78bd901487d1762b93637dad9f95))
-   change temp folder ([8d05436](https://github.com/HancomAC/HANA/commit/8d054360802a8b3685e5023a0fad67ddd264fd70))
-   do not show compile error for python & pypy ([6cf9005](https://github.com/HancomAC/HANA/commit/6cf9005c6caf9961f14f564ae05be11af1408d74))

src/app.ts
import koa from 'koa'
import Router from 'koa-router'
import webSockify from 'koa-websocket'
import send from 'koa-send'
import bodyParser from 'koa-bodyparser'

import { v4 as uuid } from 'uuid'
import { is } from 'typescript-is'
import * as os from 'os'

import { getMessageList, initWS } from './socket'
import { requestJudge } from './judge'
import { JudgeRequest } from './types/request'
import { LanguageModule, loadLanguage } from './runner/loader'
import { clearTempEnv, executeJudge, initTempEnv } from './runner/util'
import * as console from 'console'
import { getAllConfig, getConfig, setConfig } from './config'

process.on('uncaughtException', (e) => {
    console.log('uncaughtException:', e)
})

const PORT = 80
const originalRestriction = 1700000000

async function init() {
    console.log('Preloading languages...')

    const cpp = (await loadLanguage('CPP')) as LanguageModule

    if (!cpp.build) return

    console.log('Preparing sandbox...')

    const uid = uuid(),
        tempEnv = initTempEnv(uid, [
            {
                name: 'Main.cpp',
                source: `#include<iostream>
                int main() {
                    int i, j, sum = 0;
                    for (j = 0; j < 10; j++) for (i = 0; i < ${originalRestriction}; i++)
                        sum += i;
                    std::cout << sum;
                    return 0;
                }`,
            },
        ])

    console.log('Building test program...')

    console.log(await cpp.build(tempEnv, uid))

    console.log('Calculating time...')

    const { stderr } = await executeJudge(
        { uid, timeLimit: 0, memoryLimit: 1024 },
        cpp.getExecuteCommand(tempEnv, uid),
        ''
    )

    let info = '',
        err = stderr.split('\n'),
        timeUsage = 0
    while (!info.includes('|') && err.length) info = err.pop() || ''
    try {
        timeUsage = parseFloat(info.split('m ')[1].split('s')[0]) * 1000
    } catch {}

    setConfig('RunCpuLimit', (getConfig('RunCpuLimit') / 10000) * timeUsage)

    console.log(`CPU Restriction set to ${getConfig('RunCpuLimit')}%.`)

    setConfig('MultiJudgeCount', os.cpus().length - 2)

    console.log(
        `Parallel judgement count set to ${getConfig('MultiJudgeCount')}.`
    )

    console.log('Starting Server...')

    clearTempEnv(uid)

    const app = webSockify(new koa())
    const router = new Router()

    app.use(
        bodyParser({
            jsonLimit: '10mb',
        })
    )

    initWS(app)

    router.post('/judge', (ctx) => {
        try {
            const problem = {
                uid: uuid(),
                language: ctx.request.body.language,
                judgeType: ctx.request.body.judgeType,
                source: ctx.request.body.source,
                dataSet: ctx.request.body.dataSet,
                timeLimit: ctx.request.body.timeLimit,
                memoryLimit: ctx.request.body.memoryLimit,
                specialJudge: ctx.request.body.specialJudge,
            } as JudgeRequest
            if (is<JudgeRequest>(problem)) {
                requestJudge(problem)
                ctx.body = { success: true, uid: problem.uid }
            } else {
                ctx.status = 400
                ctx.body = { success: false }
            }
        } catch (e) {
            console.log(e)
            ctx.status = 400
            ctx.body = { success: false }
        }
    })

    router
        .get('/poll', (ctx) => {
            ctx.set('Cache-Control', 'no-store')
            ctx.body = { success: true, data: getMessageList() }
        })
        .get('/config', (ctx) => {
            ctx.body = { success: true, data: getAllConfig() }
        })
        .get('/test', async (ctx) => {
            await send(ctx, 'res/test.html')
        })
        .get('/res/logo', async (ctx) => {
            await send(ctx, 'res/logo.png')
        })
        .get('/res/github', async (ctx) => {
            await send(ctx, 'res/github.png')
        })
        .get('/res/docs', async (ctx) => {
            await send(ctx, 'res/docs.png')
        })
        .get('/res/test', async (ctx) => {
            await send(ctx, 'res/test.png')
        })
        .get('favicon.ico', async (ctx) => {
            await send(ctx, 'res/logo.ico')
        })
        .get('/', async (ctx) => {
            await send(ctx, 'res/index.html')
        })
        .use((ctx) => {
            ctx.redirect('http://jungol.co.kr')
        })

    app.use(router.routes()).use(router.allowedMethods())

    await app.listen(PORT)

    console.log(`Server is running on port ${PORT}.`)
}

init().catch((e) => {
    console.log(e)
})
